File: fenwick.py
DEBUG = False

def debug(msg):
    if DEBUG:
        print(msg)

class OneIndexedList:
    def __init__(self, iterable=None):
        """Initialize the list with an optional iterable."""
        self._data = list(iterable) if iterable else []

    def __getitem__(self, index):
        """Get an item using 1-based indexing."""
        if index < 1 or index > len(self._data):
            raise IndexError("Index out of range")
        return self._data[index - 1]

    def __setitem__(self, index, value):
        """Set an item using 1-based indexing."""
        if index < 1 or index > len(self._data):
            raise IndexError("Index out of range")
        self._data[index - 1] = value

    def __iter__(self):
        """Allow iteration over the list."""
        return iter(self._data)

    def __len__(self):
        """Return the number of elements in the list."""
        return len(self._data)

    def append(self, value):
        """Append a new element to the list."""
        self._data.append(value)

    def __repr__(self):
        return f"OneIndexedList({self._data})"

class Fenwick:
    def __init__(self, servers, capacity):
        self.capacity = capacity
        self.t = OneIndexedList([0]*len(servers))
        self.a = OneIndexedList(servers)
        for i, server in enumerate(servers, start=1):
            self.t[i] += server
            debug(f'\nt[{i}] = {self.t[i]-server} + {server} = {self.t[i]}')
            j = self.reverse(i)
            if j <= len(self.t):
                self.t[j] += self.t[i]
                debug(f't[{j}] = {self.t[j]-self.t[i]} + {self.t[i]} = {self.t[j]}')

    def store_data(self, size, index):
        to_store = size
        for i in range(index, min(index+5, len(self.t)+1)):
            if to_store <= 0:
                return
            stored_at_i = self.a[i]
            if to_store + stored_at_i > self.capacity:
                # There's more to store than t[i] can hold - fill it and move on
                to_store_here = self.capacity - stored_at_i
                self.update(i, to_store_here)
                to_store -= to_store_here
            else:
                # Store the rest - we can take it
                self.update(i, to_store)
                return

    def _find_deletion_index(self, t):
        # sum_before_idx is always a[1] + ... + a[idx-1]
        # idx is always the minimum possible that satisfies given info
        # mask goes 2^n -> 2^(n-1) -> ... -> 2^0
        sum_before_idx = 0
        idx = 1
        mask = 1 << (len(self.t).bit_length() - 1)
        while mask > 0:
            i = mask + idx - 1
            sum_at_i = self.t[i] + sum_before_idx if i <= len(self.t) else float('inf')
            if sum_at_i < t:
                # Go right
                idx = i + 1
                sum_before_idx = sum_at_i
            mask >>= 1
        return idx

    def delete_data(self, size, threshold):
        # 1. Find minimum i in [0..n-1] s.t. sum(i) >= t (threshold)
        # 2. Delete up to size data from t[i]
        index = self._find_deletion_index(threshold)
        stored_at_index = self.a[index]
        to_delete = min(stored_at_index, size)
        self.update(index, -to_delete)

    def query_data(self, left: int, right: int):
        '''both inclusive'''
        if right - left + 1 < right.bit_count() + (left-1).bit_count():
            # This will be faster
            return sum(self.a[i] for i in range(left, right+1))
        else:
            return self.sum(right) - self.sum(left-1)
    
    def sum(self, index):
        '''sums from indices 1 to index INCLUSIVE'''
        if index < 1:
            return 0
        if index > len(self.t):
            raise ValueError(f'Tried to get sum from 1 to {index}, which is higher than the max ({len(self.t)})')
        return self.t[index] + self.sum(index & (index-1))
    
    def reverse(self, index):
        # Add the least significant digit that's set to 1
        return index + (index & -index)
    
    def update(self, index, delta, first=True):
        debug(f'Now updating index {index} with delta {delta}')
        if index < 1 or index > len(self.t):
            return
        if first:
            self.a[index] += delta
        self.t[index] += delta
        self.update(self.reverse(index), delta, first=False)

    def __str__(self):
        return '\n'.join(f'{i}: {val}' for i, val in enumerate(self.t, start=1))

def test():
    # servers = OneIndexedList([3,7,1,4,5,4,3,1,1,2])
    servers = range(1,12)
    fenwick = Fenwick(servers, 20)
    print(fenwick)

    for i in range(1, len(servers)+1):
        print(f'sum from 1 to {i}: {fenwick.sum(i)}')

def main():
    server_count, capacity, command_count = [int(n) for n in input().strip().split()]
    servers = [int(n) for n in input().strip().split()]
    fenwick = Fenwick(servers, capacity)
    for _ in range(command_count):
        command, x, y = input().strip().split()
        x, y = int(x), int(y)
        match command:
            case 'F':
                fenwick.store_data(x, y+1)
            case 'D':
                fenwick.delete_data(x, y)
            case 'Q':
                size = fenwick.query_data(x+1, y+1)
                print(f'({x},{y}): {size}')
            case _:
                raise RuntimeError(f'Command not recognized: {command}')

if __name__ == '__main__':
    main()

File: fenwick.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <limits>
#include <cmath>
#include <stdexcept>
#include <bit>

#define DEBUG false

void debug(const std::string& msg) {
    if (DEBUG) {
        std::cout << msg << std::endl;
    }
}

class OneIndexedVector {
private:
    std::vector<size_t> data;
    
public:
    // Constructors
    OneIndexedVector(size_t size) : data(size + 1, 0) {}  // +1 for 1-based indexing
    OneIndexedVector(const std::vector<size_t>& vec) : data(vec.size() + 1, 0) {
        for (size_t i = 0; i < vec.size(); ++i) {
            data[i + 1] = vec[i];  // Fill data starting from index 1
        }
    }

    // Access operator with 1-based indexing
    size_t& operator[](size_t index) {
        if (index < 1 || index >= data.size()) {
            throw std::out_of_range("Index out of range (must be 1 to size)");
        }
        return data[index];
    }

    // Const overload for read-only access
    const size_t& operator[](size_t index) const {
        if (index < 1 || index >= data.size()) {
            throw std::out_of_range("Index out of range (must be 1 to size)");
        }
        return data[index];
    }

    // Get the size of the array (excluding the extra 0th element)
    size_t size() const {
        return data.size() - 1;
    }
};

    
class Fenwick {
private:
    size_t capacity;
    OneIndexedVector t_vector;
    OneIndexedVector a_vector;

    size_t reverse(size_t index) {
        return index + (index & -index);
    }

    void update(size_t index, int delta, bool first = true) {
        debug("Now updating index " + std::to_string(index) + " with delta " + std::to_string(delta));
        if (index < 1 || index > t_vector.size()) return;
        if (first) a_vector[index] += delta;
        t_vector[index] += delta;
        update(reverse(index), delta, false);
    }

    size_t _find_deletion_index(size_t threshold) {
        size_t sum_before_idx = 0;
        size_t idx = 1;
        for (size_t mask = 1UL << (std::bit_width(t_vector.size()) - 1); mask > 0; mask >>= 1) {
            size_t i = mask + idx - 1;
            size_t sum_at_i = (i <= t_vector.size()) ? t_vector[i] + sum_before_idx : std::numeric_limits<size_t>::max();
            if (sum_at_i < threshold) {
                idx = i + 1;
                sum_before_idx = sum_at_i;
            }
        }
        return idx;
    }

public:
    // Constructor
    Fenwick(const std::vector<size_t>& servers, size_t cap)
        : capacity(cap), 
          t_vector(servers.size()),
          a_vector(servers) {
        for (size_t i = 1; i <= servers.size(); ++i) {
            t_vector[i] += servers[i - 1];
            debug("t[" + std::to_string(i) + "] = " + std::to_string(t_vector[i] - servers[i - 1]) + " + " + std::to_string(servers[i - 1]) + " = " + std::to_string(t_vector[i]));
            size_t j = reverse(i);
            if (j <= t_vector.size()) {
                t_vector[j] += t_vector[i];
                debug("t[" + std::to_string(j) + "] = " + std::to_string(t_vector[j] - t_vector[i]) + " + " + std::to_string(t_vector[i]) + " = " + std::to_string(t_vector[j]));
            }
        }
    }

    void store_data(size_t size, size_t index) {
        size_t to_store = size;
        for (size_t i = index; i < std::min(index + 5, t_vector.size() + 1); ++i) {
            if (to_store <= 0) return;
            size_t stored_at_i = a_vector[i];
            if (to_store + stored_at_i > capacity) {
                size_t to_store_here = capacity - stored_at_i;
                update(i, to_store_here);
                to_store -= to_store_here;
            } else {
                update(i, to_store);
                return;
            }
        }
    }

    void delete_data(size_t size, size_t threshold) {
        size_t index = _find_deletion_index(threshold);
        size_t stored_at_index = a_vector[index];
        size_t to_delete = std::min(stored_at_index, size);
        update(index, -to_delete);
    }

    size_t sum(size_t index) {
        if (index < 1) return 0;
        if (index > t_vector.size()) {
            throw std::out_of_range("Tried to get sum from 1 to " + std::to_string(index) + ", which is higher than the max (" + std::to_string(t_vector.size()) + ")");
        }
        return t_vector[index] + sum(index & (index - 1));
    }

    size_t query_data(int left, int right) {
        if (right - left + 1 < __builtin_popcount(right) + __builtin_popcount(left - 1)) {
            size_t result = 0;
            for (int i = left; i <= right; ++i) {
                result += a_vector[i];
            }
            return result;
        } else {
            return sum(right) - sum(left - 1);
        }
    }
};

void main_logic() {
    size_t server_count, capacity, command_count;
    std::cin >> server_count >> capacity >> command_count;
    std::vector<size_t> servers(server_count);
    for (size_t& s : servers) std::cin >> s;
    Fenwick fenwick(servers, capacity);
    for (size_t i = 0; i < command_count; ++i) {
        char command;
        size_t x, y;
        std::cin >> command >> x >> y;
        switch (command) {
        case 'F':
            fenwick.store_data(x, y + 1);
            break;
        case 'D':
            fenwick.delete_data(x, y);
            break;
        case 'Q':
            std::cout << "(" << x << "," << y << "): " << fenwick.query_data(x+1, y+1) << std::endl;
            break;
        default:
            std::cerr << "Command not recognized: " << command << std::endl;
        }
    }
}

int main() {
    main_logic();
    return 0;
}
